jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/jdbc/JdbcAdapter.java

```diff
@@ -234,42 +234,90 @@ private static String handlePushdownRequest(final PushdownRequest request, final
+        if (columnDescription == null) {
+            sql = String.format("IMPORT FROM JDBC AT %s STATEMENT '%s'", credentials,
                     pushdownQuery.replace("'", "''"));
         } else {
-            if (JdbcAdapterProperties.userSpecifiedConnection(meta.getProperties())) {
-                credentials = JdbcAdapterProperties.getConnectionName(meta.getProperties());
-            } else {
-                credentials = "'" + connection.getAddress() + "' " + credentials;
-            }
+            sql = String.format("IMPORT INTO %s FROM JDBC AT %s STATEMENT '%s'", columnDescription, credentials,
+                    pushdownQuery.replace("'", "''"));
+        }
+        return sql;
+    }
 
-            final String columnDescription = createColumnDescription(exaMeta, meta, pushdownQuery, dialect);
-            if (columnDescription == null) {
-                sql = String.format("IMPORT FROM JDBC AT %s STATEMENT '%s'", credentials,
-                        pushdownQuery.replace("'", "''"));
-            } else {
-                sql = String.format("IMPORT INTO %s FROM JDBC AT %s STATEMENT '%s'", columnDescription, credentials,
-                        pushdownQuery.replace("'", "''"));
-            }
+    protected static String getCredentialsForJDBCImport(ExaMetadata exaMeta, SchemaMetadataInfo meta) {
+        String credentials = "";
+        if (JdbcAdapterProperties.userSpecifiedConnection(meta.getProperties())) {
+            credentials = JdbcAdapterProperties.getConnectionName(meta.getProperties());
+        } else {
+            credentials = getUserAndPasswordForImport(exaMeta, meta);
+            final ExaConnectionInformation connection = JdbcAdapterProperties.getConnectionInformation(meta.getProperties(), exaMeta);
+            credentials = "'" + connection.getAddress() + "' " + credentials;
         }
+        return credentials;
+    }
 
-        return ResponseJsonSerializer.makePushdownResponse(sql);
+    protected static String getCredentialsForORAImport(ExaMetadata exaMeta, SchemaMetadataInfo meta) {
+        String credentials = "";
+        if (!JdbcAdapterProperties.userSpecifiedConnection(meta.getProperties())) {
+            credentials = getUserAndPasswordForImport(exaMeta, meta);
+        }
+        return credentials;
+    }
+
+    protected static String getCredentialsForEXAImport(ExaMetadata exaMeta, SchemaMetadataInfo meta) {
+        return getUserAndPasswordForImport(exaMeta, meta);
+    }
+
+    private static String getUserAndPasswordForImport(ExaMetadata exaMeta, SchemaMetadataInfo meta) {
+        String credentials = "";
+        final ExaConnectionInformation connection = JdbcAdapterProperties.getConnectionInformation(meta.getProperties(),
+                    exaMeta);
+        if (connection.getUser() != null || connection.getPassword() != null) {
+            credentials = "USER '" + connection.getUser() + "' IDENTIFIED BY '" + connection.getPassword() + "'";
+        }
```

Comment:
What happens if one of the parameters is filled out and the other isn't? Shouldn't this raise an error?
I consider that a subtle security risk in case the users made a mistake here.

Reply:
This is prevented before. If you supply user without password or vice versa, the `CREATE VIRTUAL SCHEMA` statement throws an exception.